great_expectations/cli/datasource.py

```diff
@@ -1234,112 +1242,128 @@ def _get_batch_kwargs_from_generator_or_from_file_path(
     return (data_asset_name, batch_kwargs)
 
 
+def _get_default_schema(datasource):
+    inspector = sqlalchemy.inspect(datasource.engine)
+    return inspector.default_schema_name
+
+
 def _get_batch_kwargs_for_sqlalchemy_datasource(
     context, datasource_name, additional_batch_kwargs=None
 ):
-    msg_prompt_query = """
-Enter an SQL query
+    data_asset_name = None
+    sql_query = None
+    datasource = context.get_datasource(datasource_name)
+    msg_prompt_how_to_connect_to_data = """
+You have selected a datasource that is a SQL database. How would you like to specify the data?
+1. Name the table and schema I am interested in
+2. Input a custom SQL query
+3. Search for the table to add (will display all tables in database)
 """
+    default_schema = _get_default_schema(datasource)
+    temp_generator = TableBatchKwargsGenerator(name="temp", datasource=datasource)
 
-    msg_prompt_enter_data_asset_name = (
-        "\nWhich table would you like to use? (Choose one)\n"
-    )
+    while data_asset_name is None:
+        single_or_multiple_data_asset_selection = click.prompt(
+            msg_prompt_how_to_connect_to_data,
+            type=click.Choice(["1", "2", "3"]),
+            show_choices=False,
+        )
+        if single_or_multiple_data_asset_selection == "1":  # name the table and schema
+            schema_name = click.prompt(
+                "Please provide the schema name of the table (this is optional)",
+                default=default_schema,
+            )
+            table_name = click.prompt(
+                "Please provide the table name (this is required)"
+            )
+            data_asset_name = f"{schema_name}.{table_name}"
+
+        elif single_or_multiple_data_asset_selection == "2":  # SQL query
+            sql_query = click.prompt("Please provide the SQL query")
+            data_asset_name = "custom_sql_query"
+
+        elif single_or_multiple_data_asset_selection == "3":  # list it all
+            # how can we change the color here?
+            msg_prompt_warning = f"""Warning: If you have a large number of tables in your datasource, this may take a very long time. \m
+                    Would you like to continue?"""
+            confirmation = click.prompt(
+                msg_prompt_warning, type=click.Choice(["y", "n"]), show_choices=True
+            )
+            if confirmation == "y":
+                # avoid this call until necessary
+                available_data_asset_names = (
+                    temp_generator.get_available_data_asset_names()["names"]
+                )
```

Comment:
this was the function call that did the full-scan. Avoiding until the last possible moment